File: Density/script3.py
import numpy
from numpy import genfromtxt
import pandas
from scipy.spatial import distance_matrix
import hub_toolbox
from hub_toolbox.HubnessAnalysis import HubnessAnalysis


my_data = genfromtxt('20D1.csv', delimiter=',')

# calculate hubness (here, skewness of 5-occurence)
S_k, _, _ = hub_toolbox.Hubness.hubness(D=my_data, k=5, metric='distance')
# S_k = 1.5941042021329748


my_data = genfromtxt('20D2.csv', delimiter=',')

# calculate hubness (here, skewness of 5-occurence)
S_k, _, _ = hub_toolbox.Hubness.hubness(D=my_data, k=5, metric='distance')
# S_k = 2.0844290566933754


my_data = genfromtxt('20D3.csv', delimiter=',')

# calculate hubness (here, skewness of 5-occurence)
S_k, _, _ = hub_toolbox.Hubness.hubness(D=my_data, k=5, metric='distance')
# S_k = 1.2455338755901784


my_data = genfromtxt('20D4.csv', delimiter=',')

# calculate hubness (here, skewness of 5-occurence)
S_k, _, _ = hub_toolbox.Hubness.hubness(D=my_data, k=5, metric='distance')
# S_k = 2.241864419491929


my_data = genfromtxt('20D5.csv', delimiter=',')

# calculate hubness (here, skewness of 5-occurence)
S_k, _, _ = hub_toolbox.Hubness.hubness(D=my_data, k=5, metric='distance')
# S_k = 1.459787672677929


my_data = genfromtxt('20D6.csv', delimiter=',')

# calculate hubness (here, skewness of 5-occurence)
S_k, _, _ = hub_toolbox.Hubness.hubness(D=my_data, k=5, metric='distance')
# S_k = 4.275882070670827


my_data = genfromtxt('20D7.csv', delimiter=',')

# calculate hubness (here, skewness of 5-occurence)
S_k, _, _ = hub_toolbox.Hubness.hubness(D=my_data, k=5, metric='distance')
# S_k = 1.346218106509166


my_data = genfromtxt('20D8.csv', delimiter=',')

# calculate hubness (here, skewness of 5-occurence)
S_k, _, _ = hub_toolbox.Hubness.hubness(D=my_data, k=5, metric='distance')
# S_k = 1.5094910269061228


my_data = genfromtxt('20D9.csv', delimiter=',')

# calculate hubness (here, skewness of 5-occurence)
S_k, _, _ = hub_toolbox.Hubness.hubness(D=my_data, k=5, metric='distance')
# S_k = 2.1891636478365766


my_data = genfromtxt('20D10.csv', delimiter=',')

# calculate hubness (here, skewness of 5-occurence)
S_k, _, _ = hub_toolbox.Hubness.hubness(D=my_data, k=5, metric='distance')
# S_k = 1.3925336096018088

my_data = genfromtxt('20D11.csv', delimiter=',')

# calculate hubness (here, skewness of 5-occurence)
S_k, _, _ = hub_toolbox.Hubness.hubness(D=my_data, k=5, metric='distance')
# S_k = 1.5193865082408766


my_data = genfromtxt('20D12.csv', delimiter=',')

# calculate hubness (here, skewness of 5-occurence)
S_k, _, _ = hub_toolbox.Hubness.hubness(D=my_data, k=5, metric='distance')
# S_k = 4.379892856889473


my_data = genfromtxt('20D13.csv', delimiter=',')

# calculate hubness (here, skewness of 5-occurence)
S_k, _, _ = hub_toolbox.Hubness.hubness(D=my_data, k=5, metric='distance')
# S_k = 1.9983439066625033


my_data = genfromtxt('20D14.csv', delimiter=',')

# calculate hubness (here, skewness of 5-occurence)
S_k, _, _ = hub_toolbox.Hubness.hubness(D=my_data, k=5, metric='distance')
# S_k = 1.418608599846271


my_data = genfromtxt('20D15.csv', delimiter=',')

# calculate hubness (here, skewness of 5-occurence)
S_k, _, _ = hub_toolbox.Hubness.hubness(D=my_data, k=5, metric='distance')
# S_k = 1.441531169104857


my_data = genfromtxt('20D16.csv', delimiter=',')

# calculate hubness (here, skewness of 5-occurence)
S_k, _, _ = hub_toolbox.Hubness.hubness(D=my_data, k=5, metric='distance')
# S_k = 2.5854161529844366


my_data = genfromtxt('20D17.csv', delimiter=',')

# calculate hubness (here, skewness of 5-occurence)
S_k, _, _ = hub_toolbox.Hubness.hubness(D=my_data, k=5, metric='distance')
# S_k = 1.4094271234531703


my_data = genfromtxt('20D18.csv', delimiter=',')

# calculate hubness (here, skewness of 5-occurence)
S_k, _, _ = hub_toolbox.Hubness.hubness(D=my_data, k=5, metric='distance')
# S_k = 1.3327236818226298


my_data = genfromtxt('20D19.csv', delimiter=',')

# calculate hubness (here, skewness of 5-occurence)
S_k, _, _ = hub_toolbox.Hubness.hubness(D=my_data, k=5, metric='distance')
# S_k = 1.2184087479777232


my_data = genfromtxt('20D20.csv', delimiter=',')

# calculate hubness (here, skewness of 5-occurence)
S_k, _, _ = hub_toolbox.Hubness.hubness(D=my_data, k=5, metric='distance')
# S_k = 1.6662435829103943

File: Density/script3.R
library(MVA)

####################

x <- matrix(rnorm(2000, 50, 0.05), ncol=20)
x2 <- data.frame(x)

x3 <- as.list(dist(x, diag = FALSE, upper = FALSE))
x4 <- as.data.frame(x3)
rowMeans(x4) # 0.3008012

m <- as.matrix(dist(x))
m2 <- as.data.frame(m)

write.csv(m2, "20D1.csv", row.names = FALSE)

####################

x <- matrix(rnorm(2000, 50, 0.1), ncol=20)
x2 <- data.frame(x)

x3 <- as.list(dist(x, diag = FALSE, upper = FALSE))
x4 <- as.data.frame(x3)
rowMeans(x4) # 0.6357493

m <- as.matrix(dist(x))
m2 <- as.data.frame(m)

write.csv(m2,"20D2.csv", row.names = FALSE)

####################

x <- matrix(rnorm(2000, 50, 0.2), ncol=20)
x2 <- data.frame(x)

x3 <- as.list(dist(x, diag = FALSE, upper = FALSE))
x4 <- as.data.frame(x3)
rowMeans(x4) # 1.235096

m <- as.matrix(dist(x))
m2 <- as.data.frame(m)

write.csv(m2, "20D3.csv", row.names = FALSE)

####################

x <- matrix(rnorm(2000, 50, 0.25), ncol=20)
x2 <- data.frame(x)

x3 <- as.list(dist(x, diag = FALSE, upper = FALSE))
x4 <- as.data.frame(x3)
rowMeans(x4) # 1.511662

m <- as.matrix(dist(x))
m2 <- as.data.frame(m)

write.csv(m2, "20D4.csv", row.names = FALSE)

####################

x <- matrix(rnorm(2000, 50, 0.3), ncol=20)
x2 <- data.frame(x)

x3 <- as.list(dist(x, diag = FALSE, upper = FALSE))
x4 <- as.data.frame(x3)
rowMeans(x4) # 1.90205

m <- as.matrix(dist(x))
m2 <- as.data.frame(m)

write.csv(m2, "20D5.csv", row.names = FALSE)

####################

x <- matrix(rnorm(2000, 50, 0.5), ncol=20)
x2 <- data.frame(x)

x3 <- as.list(dist(x, diag = FALSE, upper = FALSE))
x4 <- as.data.frame(x3)
rowMeans(x4) # 3.138649

m <- as.matrix(dist(x))
m2 <- as.data.frame(m)

write.csv(m2, "20D6.csv", row.names = FALSE)

####################

x <- matrix(rnorm(2000, 50, 0.75), ncol=20)
x2 <- data.frame(x)

x3 <- as.list(dist(x, diag = FALSE, upper = FALSE))
x4 <- as.data.frame(x3)
rowMeans(x4) # 4.568908

m <- as.matrix(dist(x))
m2 <- as.data.frame(m)

write.csv(m2, "20D7.csv", row.names = FALSE)

####################

x <- matrix(rnorm(2000, 50, 0.85), ncol=20)
x2 <- data.frame(x)

x3 <- as.list(dist(x, diag = FALSE, upper = FALSE))
x4 <- as.data.frame(x3)
rowMeans(x4) # 5.140225

m <- as.matrix(dist(x))
m2 <- as.data.frame(m)

write.csv(m2, "20D8.csv", row.names = FALSE)

####################

x <- matrix(rnorm(2000, 50, 1), ncol=20)
x2 <- data.frame(x)

x3 <- as.list(dist(x, diag = FALSE, upper = FALSE))
x4 <- as.data.frame(x3)
rowMeans(x4) # 6.270466

m <- as.matrix(dist(x))
m2 <- as.data.frame(m)

write.csv(m2, "20D9.csv", row.names = FALSE)

####################

x <- matrix(rnorm(2000, 50, 1.25), ncol=20)
x2 <- data.frame(x)

x3 <- as.list(dist(x, diag = FALSE, upper = FALSE))
x4 <- as.data.frame(x3)
rowMeans(x4) # 7.795185

m <- as.matrix(dist(x))
m2 <- as.data.frame(m)

write.csv(m2, "20D10.csv", row.names = FALSE)

####################

x <- matrix(rnorm(2000, 50, 1.5), ncol=20)
x2 <- data.frame(x)

x3 <- as.list(dist(x, diag = FALSE, upper = FALSE))
x4 <- as.data.frame(x3)
rowMeans(x4) # 9.405675

m <- as.matrix(dist(x))
m2 <- as.data.frame(m)

write.csv(m2, "20D11.csv", row.names = FALSE)

####################

x <- matrix(rnorm(2000, 50, 1.75), ncol=20)
x2 <- data.frame(x)

x3 <- as.list(dist(x, diag = FALSE, upper = FALSE))
x4 <- as.data.frame(x3)
rowMeans(x4) # 11.06962

m <- as.matrix(dist(x))
m2 <- as.data.frame(m)

write.csv(m2, "20D12.csv", row.names = FALSE)

####################

x <- matrix(rnorm(2000, 50, 2), ncol=20)
x2 <- data.frame(x)

x3 <- as.list(dist(x, diag = FALSE, upper = FALSE))
x4 <- as.data.frame(x3)
rowMeans(x4) # 12.56801

m <- as.matrix(dist(x))
m2 <- as.data.frame(m)

write.csv(m2, "20D13.csv", row.names = FALSE)

####################

x <- matrix(rnorm(2000, 50, 3), ncol=20)
x2 <- data.frame(x)

x3 <- as.list(dist(x, diag = FALSE, upper = FALSE))
x4 <- as.data.frame(x3)
rowMeans(x4) # 19.00258

m <- as.matrix(dist(x))
m2 <- as.data.frame(m)

write.csv(m2, "20D14.csv", row.names = FALSE)

####################

x <- matrix(rnorm(2000, 50, 5), ncol=20)
x2 <- data.frame(x)

x3 <- as.list(dist(x, diag = FALSE, upper = FALSE))
x4 <- as.data.frame(x3)
rowMeans(x4) # 31.70717

m <- as.matrix(dist(x))
m2 <- as.data.frame(m)

write.csv(m2, "20D15.csv", row.names = FALSE)

####################

x <- matrix(rnorm(2000, 50, 10), ncol=20)
x2 <- data.frame(x)

x3 <- as.list(dist(x, diag = FALSE, upper = FALSE))
x4 <- as.data.frame(x3)
rowMeans(x4) # 61.86532

m <- as.matrix(dist(x))
m2 <- as.data.frame(m)

write.csv(m2, "20D16.csv", row.names = FALSE)

####################

x <- matrix(rnorm(2000, 50, 15), ncol=20)
x2 <- data.frame(x)

x3 <- as.list(dist(x, diag = FALSE, upper = FALSE))
x4 <- as.data.frame(x3)
rowMeans(x4) # 94.79673

m <- as.matrix(dist(x))
m2 <- as.data.frame(m)

write.csv(m2, "20D17.csv", row.names = FALSE)

####################

x <- matrix(rnorm(2000, 50, 20), ncol=20)
x2 <- data.frame(x)

x3 <- as.list(dist(x, diag = FALSE, upper = FALSE))
x4 <- as.data.frame(x3)
rowMeans(x4) # 124.4367

m <- as.matrix(dist(x))
m2 <- as.data.frame(m)

write.csv(m2, "20D18.csv", row.names = FALSE)

####################

x <- matrix(rnorm(2000, 50, 25), ncol=20)
x2 <- data.frame(x)

x3 <- as.list(dist(x, diag = FALSE, upper = FALSE))
x4 <- as.data.frame(x3)
rowMeans(x4) # 159.945

m <- as.matrix(dist(x))
m2 <- as.data.frame(m)

write.csv(m2, "20D19.csv", row.names = FALSE)

####################

x <- matrix(rnorm(2000, 50, 30), ncol=20)
x2 <- data.frame(x)

x3 <- as.list(dist(x, diag = FALSE, upper = FALSE))
x4 <- as.data.frame(x3)
rowMeans(x4) # 184.376

m <- as.matrix(dist(x))
m2 <- as.data.frame(m)

write.csv(m2, "20D20.csv", row.names = FALSE)



# newdata <- mydata[1:15,]

mydata <- read.csv("20Ddata.csv")

library("ggpubr")
ggscatter(newdata, x = "hubness", y = "density", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          xlab = "Hubness", ylab = "Density")